Do not run dependent jobs when parent is canceled

fredsod opened the pull request:
I ran into a problem when we canceled a running parent job and the dependent jobs were executed.

When rq cancel a running job the status is changed in the database but when rq tries to enqueue the dependent jobs it seems like it does not fetch the new value from the database.

I do not know that much about the internals of RQ, could be better ways to fix this.

Probably fixes #1902

selwin commented:
Hey I think a better fix would be what's suggested by @parthvnp here: https://github.com/rq/rq/issues/1902#issuecomment-1543324530

Attaching a passing test like what @amks1 described here is also necessary https://github.com/rq/rq/issues/1902#issue-1696852254

fredsod commented:
Hi, I have tried the fix proposed in #1902 but I can't get it to work for my test cases. It could be that the fix it's needed if you try to enqueue a job to an already canceled job. But I think there are multiple problems when trying to cancel both running and queued jobs.

I have updated the PR with two more test cases that fail on the master branch, I also added a check if the dependent job is canceled when a finished job tries to enqueue its dependent jobs.

selwin commented on tests/test_worker.py:
```diff
@@ -594,6 +595,91 @@ def test_deleted_jobs_arent_executed(self):
+        status_thread.join()
+
+        self.assertNotEqual(parent_job.result, None)
+        self.assertEqual(job.get_status(), "deferred")
```
Mind using status enums for comparison like so? `self.assertEqual(job.get_status(), JobStatus.DEFERRED)`

selwin commented on tests/test_worker.py:
```diff
@@ -594,6 +595,91 @@ def test_deleted_jobs_arent_executed(self):
+            job.cancel()
+            return
+
+        q = Queue(
+            "low",
+        )
```
This should be one line, and please format this patch with black.

selwin commented:
> Hi, I have tried the fix proposed in https://github.com/rq/rq/issues/1902 but I can't get it to work for my test cases. It could be that the fix it's needed if you try to enqueue a job to an already canceled job. But I think there are multiple problems when trying to cancel both running and queued jobs.

You're right, the bug you described happens because the parent job itself doesn't know it has been canceled in the background. This PR looks good, but I left some comments just to make sure the patch is stylistically consistent with the rest of the code base.
